src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java

```diff
@@ -263,12 +273,30 @@ public boolean hasMapping(String index) {
    * document id when either the DLQ is configured or
    * {@link ElasticsearchSinkConnectorConfig#IGNORE_KEY_CONFIG} is set to <code>false</code> because
    * they require the use of a map keyed by document id.
+   * <br>
+   * This call is usually asynchronous, but can block if:
+   * <ul>
+   *   <li>A new batch is finished (e.g. max batch size has been reached) and
+   *    the overall number of threads (max in flight requests) are in use.</li>
+   *   <li>The maximum number of buffered records have been reached</li>
+   * </ul>
    *
    * @param record the record to index
    * @param request the associated request to send
    * @throws ConnectException if one of the requests failed
    */
-  public void index(SinkRecord record, DocWriteRequest<?> request) {
+  public void index(SinkRecord record, DocWriteRequest<?> request, OffsetState offsetState) {
+    throwIfFailed();
+
+    // TODO should we just pause partitions instead of blocking and failing the connector?
```

Comment:
Are we or are we not pausing partitions?

Reply:
Nevermind. I totally missed where the `PartitionPauser` logic is.

Reply:
We are pausing partitions in case we reach the limit of offset entries.

This is another scenario, where currently we block for some time and fail the connector if the overall number of in-flight records reaches the predefined limit. 
I marked this as a TODO to limit the scope of this PR.